### ISSUE TITLE
Fix PromiseConstructor interface

### ISSUE THREAD

**Issue by falsandtru** [microsoft/TypeScript#10699]
**TypeScript Version:**  master

``` diff
interface PromiseConstructor {
-    reject(reason: any): Promise<never>;
-    reject<T>(reason: any): Promise<T>;
+    reject(reason?: any): Promise<never>;
}
```

**Comment by mhegazy** [comment]
The second overload is handy for casts. it does not get picked up aromatically, it is only used if someone passes an explicit type argument. so i would keep that. 

changing the argument to be optional seems correct. a PR would be appreciated.

**Comment by falsandtru** [comment]
`Promise.reject` never changes the contextual type of Promise. So I think the fix your thoughts would be this. However, it is buggy. Anyone cannot trust a `reason` parameter type. Do you really need that?

```diff
interface PromiseConstructor {
-    reject(reason: any): Promise<never>;
-    reject<T>(reason: any): Promise<T>;
+    reject(reason?: any): Promise<never>;
+    reject<T>(reason: T): Promise<never>;
}
```

Add:

Could you accept the following fix?

```diff
interface PromiseConstructor {
-    new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
+    new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
}
```

for the current strict typings of `Promise.resolve`.

```ts
    resolve<T>(value: T | PromiseLike<T>): Promise<T>;
    resolve(): Promise<void>;
```

**Comment by kitsonk** [comment]
The cast is useful if you had something like this:

```typescript
function looksLikeAsync(): Promise<string> {
    if (someCondition) {
        return Promise.resolve('foo');
    else {
        return Promise.reject<string>(new Error('I hate kittens'));
    }
}

looksLikeAsync()
    .then((str) => { /* do something */ }, (err) => { /* handle error */ });
```

Previously though `Promise<void>` was not assignable to `Promise<string>` but `Promise<never>` is always assignable to `Promise<string>`, though I suspect changing the typing and taking away the generic would break other peoples code who previously had to cast before `never` was a thing.

**Comment by falsandtru** [comment]
I believe its typing is verbose.

**Comment by kitsonk** [comment]
> I believe its typing is verbose.

But it is the only way to accomplish the use case I provided.  There are other ways to cast, but I am fairly certain people have relied upon that and would be a breaking change.  Breaking other people's code for your personal sense of verbosity is cruel when it has no direct impact on your code or usage of TypeScript.

**Comment by falsandtru** [comment]
For example, the following case returns the unexpected type. This is very bad.

```ts
// An expected return type is Promise<void>, actually it returns Promise<string | void>
function f() {
  return Promise.resolve()
    .then(() => Promise.resolve(), () => Promise.reject<string>(''));
}
```

**Comment by mhegazy** [comment]
you wrote `<string>`. not sure what you expected, and not sure what the compiler could have done to save you from your self, you could as easily write `<string>Promise.reject('')`

**Comment by falsandtru** [comment]
`<string>Promise.reject('')` is not compilable. Can you provide the case your thoughts?

**Comment by mhegazy** [comment]
sorry, meant `Promise<string>`. so:

```ts
function f() {
    return Promise.resolve()
        .then(() => Promise.resolve(), () => <Promise<string>>Promise.reject(''));
}
```

is basically the same as:

```ts
// An expected return type is Promise<void>, actually it returns Promise<string | void>
function f() {
  return Promise.resolve()
    .then(() => Promise.resolve(), () => Promise.reject<string>(''));
}
```

except that the later is easier to write and nicer to read.

**Comment by falsandtru** [comment]
Why you make `Promise<string | void>`? Its type is wrong. I think TypeScript shouldn't provides the way to make incorrect types. So `Promise.reject` should always return `Promise<never>` to make a correct contextual type.

**Comment by mhegazy** [comment]
that is what `Promise.reject(value)` returns.

if some one chose to pass an explicit type argument, e.g. `Promise.reject<any>(value)` i do not see why we should stop them. again this is identical to `<Promise<any>> Promise.reject(value)`, just easier to write.

**Comment by falsandtru** [comment]
`<Promise<number>>Promise.reject(0)` and `<Promise<number>>Promise.resolve(0)` have different effects, but they have a same type. `<Promise<number>>Promise.reject(0)` never make `Promise<number>` by themselves.

```ts
Promise.reject<string>(0)
  .catch(() => <Promise<number>>Promise.resolve(0))
  .then(n => console.log(n)); // 0

Promise.reject<string>(0)
  .catch(() => <Promise<number>>Promise.reject(0))
  .then(n => console.log(n)); // Uncaught (in promise) 0
```

So actually, your type casting means `<Promise<number>><Promise<never>>Promise.reject(0)`. So the minimum type of `Promise.reject(0)` is `Promise<never>`, `<Promise<number>>` is verbose and has no reason in that case. Signatures should defined by minimum types.

Why you need to make `Promise<T>` type by `Promise.reject<T>(reason: any)` even though `Promise<never>` can assign any other `Promise<T>`? And, why you need to hide that difference of effects? `Promise<never>` would help to know the actual behavior (effect). Standard definitions should help that standard understanding.

**Comment by mhegazy** [comment]
just to be clear, this bug is for marking the parameter as optional. removing the overload is not desired.

**Comment by mhegazy** [comment]
not sure i understand what you mean.

**Comment by falsandtru** [comment]
i want to modify the commit in #10819. so if you don't reopen that pr, i need to open the new pr.

**Comment by jablko** [comment]
I believe this issue was fixed in #22772 and can now be closed?

**Comment by falsandtru** [comment]
That PR is reverted but anyway the issue is fixed now.
